iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl

```diff
@@ -65,122 +88,132 @@ inline bool SpscSofi<ValueType, CapacityValue>::setCapacity(const uint64_t newSi
 template <class ValueType, uint64_t CapacityValue>
 inline bool SpscSofi<ValueType, CapacityValue>::empty() const noexcept
 {
-    uint64_t currentReadPosition{0};
-    bool isEmpty{false};
-
-    do
-    {
-        /// @todo iox-#1695 read before write since the writer increments the aba counter!!!
-        /// @todo iox-#1695 write doc with example!!!
-        currentReadPosition = m_readPosition.load(std::memory_order_acquire);
-        uint64_t currentWritePosition = m_writePosition.load(std::memory_order_acquire);
-
-        isEmpty = (currentWritePosition == currentReadPosition);
-        // we need compare without exchange
-    } while (!(currentReadPosition == m_readPosition.load(std::memory_order_acquire)));
-
-    return isEmpty;
+    auto [readPost, writePos] = getReadWritePositions();
+    return readPost == writePos;
 }
 
 template <class ValueType, uint64_t CapacityValue>
 inline bool SpscSofi<ValueType, CapacityValue>::pop(ValueType& valueOut) noexcept
 {
-    return popIf(valueOut, [](ValueType) { return true; });
-}
-
-template <class ValueType, uint64_t CapacityValue>
-template <typename Verificator_T>
-inline bool SpscSofi<ValueType, CapacityValue>::popIf(ValueType& valueOut, const Verificator_T& verificator) noexcept
-{
-    uint64_t currentReadPosition = m_readPosition.load(std::memory_order_acquire);
-    uint64_t nextReadPosition{0};
+    // Memory order relaxed is enough since:
+    // - synchronization is not needed with m_readPosition
+    // - operation reordering: cannot compile if moves down
+    uint64_t currentReadPos = m_readPosition.load(std::memory_order_relaxed);
 
-    bool popWasSuccessful{true};
     do
     {
-        if (currentReadPosition == m_writePosition.load(std::memory_order_acquire))
+        // SYNC POINT READ: m_data
+        // See explanation of the corresponding synchronization point in push()
+        if (currentReadPos == m_writePosition.load(std::memory_order_acquire))
         {
-            nextReadPosition = currentReadPosition;
-            popWasSuccessful = false;
+            return false;
+            // We don't need to check if read has changed, as it is enough to know that the empty state
+            // was valid in the past. The same race can also happen after the while loop and before the
+            // return operation
         }
```

Comment:
Just for clarification, here some ASCII art
```
Start condition
- SOFI with an internal capacity of 3
- read position = 301
- read index = 1
- write position = 303
- write index = 0
                  |                 |
+-----+-----+-----+-----+-----+-----+-----+
|  -  |  B  |  C  |  -  |  -  |  -  |  -  | data
+-----+-----+-----+-----+-----+-----+-----+
|  0  |  1  |  2  |  0  |  1  |  2  |  0  | index
+-----+-----+-----+-----+-----+-----+-----+
| 300 | 301 | 302 | 303 | 304 | 305 | 306 | position
+-----+-----+-----+-----+-----+-----+-----+
         r        |  w              |

Now the following happens
- consumer: m_writePosition.load(std::memory_order_acquire) -> 303
- producer: push two times -> read position move to 303

                  |                 |
+-----+-----+-----+-----+-----+-----+-----+
|  -  |  -  |  -  |  D  |  E  |  -  |  -  | data
+-----+-----+-----+-----+-----+-----+-----+
|  0  |  1  |  2  |  0  |  1  |  2  |  0  | index
+-----+-----+-----+-----+-----+-----+-----+
| 300 | 301 | 302 | 303 | 304 | 305 | 306 | position
+-----+-----+-----+-----+-----+-----+-----+
                  |  r           w  |
                  
- consumer: m_readPosition.load(std::memory_order_relaxed) -> 303 is stored in currentReadPos
    -> if (303 == 303)
    -> return false
```

Reply:
You have a point that this comment is probably wrong:

>  operation reordering: cannot compile if moves down

About "resulting in the hallucination that the SOFI is empty although it is full": isn't something that is inherent to a concurrent queue, i.e. you call push and then pop, pop doesn't return anything. You call again pop and this time, it doesn't return a value.

Reply:
> isn't something that is inherent to a concurrent queue, i.e. you call push and then pop, pop doesn't return anything. You call again pop and this time, it doesn't return a value.

I guess you refer to this comment https://github.com/eclipse-iceoryx/iceoryx/pull/2168#discussion_r1480053367

It is different in this case. Just assume what would happen on the consumer side if there wasn't a concurrent push on the producer side. In the linked comment, the result would be the same, an empty pop. The situation here is different. In this scenario the pop would not be empty if there were no concurrent push, which means the queue was not empty an therefore it is a bug and not one of these inherent races. These inherent races affect only future events, not past ones.

You can also imagine it as a queue with a capacity of 1 (internally it will be two due to the addon). Let's further assume the producer threads pushes ten times as often as the consumer thread pops. I think we can agree that there should always be data in the queue when the pop occurs. Which this hallucination it could theoretically look like the queue is empty every time the consumer thread tries to pop. Granted, it is not likely to happen all the time (it's a race after all) but is shows that the algorithm is broken.

Reply:
Ok fixed